cps/main.c

```diff
@@ -16,11 +16,529 @@
+			rte_pktmbuf_free(bufs[i]);
+	}
+}
+
+static void
+process_egress(struct gatekeeper_if *iface, struct rte_kni *kni,
+	uint16_t tx_queue)
+{
+	struct rte_mbuf *bufs[GATEKEEPER_MAX_PKT_BURST];
+	uint16_t num_rx = rte_kni_rx_burst(kni, bufs, GATEKEEPER_MAX_PKT_BURST);
+	unsigned int num_tx = rte_eth_tx_burst(iface->id, tx_queue,
+		bufs, num_rx);
+
+	/*
+	 * TODO Forward BGP, respond to ARP or ND requests with resolution,
+	 * and drop everything else.
+	 */
+
+	if (unlikely(num_tx < num_rx)) {
+		uint16_t i;
+		for (i = 0; i < num_rx - num_tx; i++)
+			rte_pktmbuf_free(bufs[i]);
+	}
+}
+
+static int
+cps_proc(void *arg)
+{
+	struct cps_config *cps_conf = (struct cps_config *)arg;
+	struct net_config *net_conf = cps_conf->net;
+
+	struct gatekeeper_if *front_iface = &net_conf->front;
+	struct gatekeeper_if *back_iface = &net_conf->back;
+	struct rte_kni *front_kni = cps_conf->front_kni;
+	struct rte_kni *back_kni = cps_conf->back_kni;
+
+	RTE_LOG(NOTICE, GATEKEEPER,
+		"cps: the CPS block is running at lcore = %u\n",
+		cps_conf->lcore_id);
+
+	while (likely(!exiting)) {
+		/*
+		 * Read in IPv4 BGP packets that arrive directly
+		 * on the Gatekeeper interfaces.
+		 */
+		process_ingress(front_iface, front_kni,
+			cps_conf->rx_queue_front);
+		if (net_conf->back_iface_enabled)
+			process_ingress(back_iface, back_kni,
+				cps_conf->rx_queue_back);
+
+		/*
+		 * Process any requests made to the CPS block, including
+		 * IPv6 BGP packets that arrived via an ACL.
+		 */
+		process_reqs(cps_conf);
+
+		/*
+		 * Read in packets from KNI interfaces, and
+		 * transmit to respective Gatekeeper interfaces.
+		 */
+		process_egress(front_iface, front_kni,
+			cps_conf->tx_queue_front);
+		if (net_conf->back_iface_enabled)
+			process_egress(back_iface, back_kni,
+				cps_conf->tx_queue_back);
+
+		/* TODO Get route updates from kernel. */
+	}
+
+	RTE_LOG(NOTICE, GATEKEEPER,
+		"cps: the CPS block at lcore = %u is exiting\n",
+		cps_conf->lcore_id);
+
+	return cleanup_cps();
+}
+
+static int
+cps_req(enum cps_req_ty ty, void *req_arg)
+{
+	struct cps_config *cps_conf = get_cps_conf();
+	struct cps_request *req = mb_alloc_entry(&cps_conf->requests);
+	int ret;
+
+	if (req == NULL) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"cps: allocation for request of type %d failed", ty);
+		return -1;
+	}
+
+	req->ty = ty;
+
+	switch (ty) {
+	case CPS_REQ_BGP:
+		req->u.bgp = *(struct cps_bgp_req *)req_arg;
+		break;
+	default:
+		mb_free_entry(&cps_conf->requests, req);
+		RTE_LOG(ERR, GATEKEEPER,
+			"cps: unknown request type %d failed", ty);
+		return -1;
+	}
+
+	ret = mb_send_entry(&cps_conf->requests, req);
+	if (ret < 0)
+		return ret;
+
+	return 0;
+}
+
+static int
+submit_cps(struct rte_mbuf **pkts, int num_pkts, struct gatekeeper_if *iface)
+{
+	struct cps_config *cps_conf = get_cps_conf();
+	struct cps_bgp_req bgp_req = {
+		.num_pkts = num_pkts,
+		.kni = iface == &cps_conf->net->front
+			? cps_conf->front_kni
+			: cps_conf->back_kni,
+	};
```

Comment:
I added this in, but still kept the format of the CPS request overall (using a union, similar to the LLS request) in case we need to add more types of requests in the future, such as from the dynamic configuration.